packages/core/test/predictableExec.test.ts

```diff
@@ -584,4 +585,25 @@ describe('predictableExec', () => {
 
     expect(actual).toEqual(['stop 1', 'start 2']);
   });
+
+  // TODO: this might be tricky in v5 because this currently relies on `.exec` property being mutated to capture the context values in a closure
```

Comment:
I don't think this will be tricky, especially if we execute actions eagerly in v5, as you suggested?

Reply:
That's why currently there's the notion of `preInitialState` in v5, which represents the initial state and actions _without_ any assignments performed nor actions executed. 

In my view, this is what `machine.initialState` (the pure getter) should be - state without actions executed. Then when the machine starts (is interpreted), the actions should be executed.

So in this test, `machine.initialState` would have context `{ count: 0 }` since no actions were executed yet, but the read state from `interpreter.subscribe(state => ...)` would have `{ count: 2 }` since the state after the initial state has those actions executed.

This would also be what the input RFC would solve, since there would be no need for initial `assign` actions.

Half-baked thoughts, just thinking out loud